src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SslOverTdsStream.cs

```diff
@@ -0,0 +1,212 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO;
+
+namespace System.Data.SqlClient.SNI
+{
+    /// <summary>
+    /// SSL encapsulated over TDS transport. During SSL handshake, SSL packets are
+    /// transported in TDS packet type 0x12. Once SSL handshake has completed, SSL
+    /// packets are sent transparently.
+    /// </summary>
+    internal sealed class SslOverTdsStream : Stream
+    {
+        private Stream _stream;
+        private bool _encapsulate;
+        private int _packetBytes = 0;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="stream">Underlying stream</param>
+        public SslOverTdsStream(Stream stream)
+        {
+            _stream = stream;
+            _encapsulate = true;
+        }
+
+        /// <summary>
+        /// Finish SSL handshake. Stop encapsulating in TDS.
+        /// </summary>
+        public void FinishHandshake()
+        {
+            _encapsulate = false;
+        }
+
+        /// <summary>
+        /// Read buffer
+        /// </summary>
+        /// <param name="buffer">Buffer</param>
+        /// <param name="offset">Offset</param>
+        /// <param name="count">Byte count</param>
+        /// <returns>Bytes read</returns>
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            int readBytes;
+            byte[] scratch = new byte[count < 8 ? 8 : count];
+
+            if (_encapsulate)
+            {
+                if (_packetBytes == 0)
+                {
+                    readBytes = _stream.Read(scratch, 0, 8);
+                    _packetBytes = scratch[2] * 256;
+                    _packetBytes += scratch[3];
+                    _packetBytes -= 8;
+                }
+
+                if (count > _packetBytes)
+                {
+                    count = _packetBytes;
+                }
+            }
+
+            readBytes = _stream.Read(scratch, 0, count);
```

Comment:
Do we completely control the stream that's being used here?  I'm just wondering if we need to be concerned with bad/maliicious streams returning bad data, e.g. a readBytes value that's negative or that's larger than count?